typing/value_rec_check.ml

```diff
@@ -189,28 +208,130 @@ let classify_expression : Typedtree.expression -> sd =
+                | Tstr_module ({mb_presence=Mp_present} as mb) ->
+                  begin match mb.mb_id with
+                  | None -> acc
+                  | Some _ -> acc + 1
+                  end
+                | Tstr_recmodule bindings ->
+                  let ids = (List.filter_map (fun mb -> mb.mb_id) bindings) in
+                  acc + List.length ids
+                | Tstr_class cl_list ->
+                  acc + List.length cl_list
+                | Tstr_include incl ->
+                  let ids = bound_value_identifiers incl.incl_type in
+                  acc + List.length ids
+                | Tstr_open od ->
+                  let ids = bound_value_identifiers od.open_bound_items in
+                  acc + List.length ids
+                | Tstr_primitive _
+                | Tstr_type _
+                | Tstr_eval _
+                | Tstr_module ({mb_presence=Mp_absent})
+                | Tstr_modtype _
+                | Tstr_class_type _
+                | Tstr_attribute _ ->
+                  acc
+            in
+            size_of_structure_items acc rem
+          in
+          size_of_structure_items 0 s.str_items
+        in
+        let rec classify_mexp mexp : sd =
+          match mexp.mod_desc with
+          | Tmod_ident _ ->
+            Not_recursive
+          | Tmod_structure s ->
+            Static (Regular_block (size_of_structure s))
+          | Tmod_functor _ ->
+            Static Function
+          | Tmod_apply _ ->
+            Not_recursive
+          | Tmod_apply_unit _ ->
+            Not_recursive
+          | Tmod_constraint (mexp, _, _, coe) ->
+            begin match coe with
+            | Tcoerce_none -> classify_mexp mexp
+            | Tcoerce_structure (pos_cc_list, _) ->
+                Static (Regular_block (List.length pos_cc_list))
+            | Tcoerce_functor _ -> Static Function
+            | Tcoerce_primitive _ ->
+              Misc.fatal_error "letrec: primitive coercion on a module"
+            | Tcoerce_alias _ ->
+              Misc.fatal_error "letrec: alias coercion on a module"
+            end
+          | Tmod_unpack (e, _) ->
+            classify_expression env e
+        in
+        classify_mexp mexp
+    | Texp_function _ ->
+        Static Function
+    | Texp_lazy e ->
+      (* The code below was copied (in part) from translcore.ml *)
+      begin match Typeopt.classify_lazy_argument e with
+      | `Constant_or_function ->
+        (* A constant expr (of type <> float if [Config.flat_float_array] is
+           true) gets compiled as itself. *)
+          classify_expression env e
+      | `Float_that_cannot_be_shortcut
+      | `Identifier `Forward_value ->
+          (* Forward blocks *)
+          Static (Regular_block 1)
+      | `Identifier `Other ->
+          classify_expression env e
+      | `Other ->
+          (* other cases compile to a lazy block holding a function *)
+          Static (Regular_block 1)
```

Comment:
I feel a bit nervous about those magic constants in the module. For example, we discussed changing the layout of lazy thunks to get two fields.

It would be nice to define those in lambda.mli:

```ocaml
val lazy_block_size : int
val variant_block_size : arity:int -> int
val extension_constructor_size : int
...
```

Reply:
Maybe we could already do the split in the current PR? I am hesitant about merging the current PR in a state that we don't find so satisfying, because the next one is larger and more invasive (it does change the generated code in a noticeable way) so it may take longer to get in.

Reply:
Fun fact: we are already in a slightly unstable state regarding dependencies. `Value_rec_check` (previously `Rec_check`) depends on `Lambda`, but is linked before it.
It all works out because we actually only depend on constructors for pattern matching, so no runtime dependencies. We still have a dependency from `value_rec_check.cmx` to `lambda.cmx` in `.depend`, but that's "fine" again because the dependency doesn't actually get registered in the `.cmx` file.
On the other hand, introduing a real dependency from `Value_rec_check` to `Lambda`, even through indirections, will inevitably make the linker complain.

So I see three main solutions. One is to revert #12586 so that we can specify a link order that follows actual dependencies, not just directories. That would probably make @shindere sad though, so I don't advise it.
The second solution is to tweak our build system so that the dependencies and the link order agree with each other. This could mean moving files from one directory to another, splitting definitions into separate files, or adapting the Makefile. I'll let @shindere comment on which one is preferred.
The last solution is to do nothing, and give up on proper splitting. This is the most convenient for me because I don't have to do anything, but I agree it's not very satisfying.

I am probably overlooking other possible solutions, feel free to propose alternatives.

Reply:
In this context, I agree with delaying the split.

I think that it is wise to avoid dependencies from the type-checking phase to the code production phase. This is not just a matter of build system, it makes the design simpler.

One approach could be to split the recursive_kind computation in two passes:
- one running during type-checking that merely classifies value among Static, Not_recursive (maybe class?), without distinguishing between Constant and Static or trying to compute a size
- one running during the typedtree->lambda compilation, that refines the first classification by distinguishing constants, computing sizes, etc.

The question then would be how to formulate this to avoid redundant computations (and a risk of mismatch) between the two phases. My suggestion would be to include in the first phase any part of the typedtree that we know will be relevant to refine the classification in the second class: `Static of rec_shape`, where `rec_shape` is a tailored AST supporting all interesting rec_check cases.

Reply:
On paper, that looks very nice. It avoids estimating sizes on the typedtree, and on lambda sizes are already computed. Any mismatch will result in a fatal error at compilation (if we cannot compute the size of a `Static` binding), or "valid" recursive definitions getting rejected (if we classify them as `Not_recursive` even though they correspond to a simple allocation).

I don't think we actually need to store a shape; all these weird cases get resolved during translation to lambda anyway. However, we will likely need to keep the lazy classification code and the traversal of module expressions (without the size computation) in the typedtree version of `Rec_check`.

I'll try to implement it and see if it actually works.

Reply:
I have implemented that on the main PR (#12596). I will now try to reduce this PR to the bugfixes only (finer grained classification for modules and lazy, to fix the two outstanding bugs).

Reply:
I only skimmed the patch but I like where this is going, it looks simpler. (In particular, we don't have to second-guess what will become an Lconst anymore.)